### PR TITLE
[ADVAPP-635]: Update AI chat cloning to resolve message limitation

### PR DESCRIPTION
**Pull request by danharrin**
### Ticket(s) or GitHub Issue

- https://canyongbs.atlassian.net/browse/ADVAPP-635

### Technical Description

A thread can be locked, when it has a `locked_at` set to a timestamp. When reinitialising a thread, it becomes locked. When a user sends or retries a message with a locked thread, they get an error message immediately that automatically updates when the thread becomes unlocked, allowing them to retry their message. The locked status of a thread gets checked every 10 seconds while the error message is present.

From my investigation we don't need to send test messages or runs when reinitialising a thread.  When a thread has over 32 messages, we send the first 32 as part of the thread creation request and then send messages individually after that.

### Screenshots (if appropriate)

Screenshot 2024-06-10 at 14 56 47

Screenshot 2024-06-10 at 14 57 12

### CODE REVIEW

**Review by mxm1070 — CHANGES_REQUESTED** [review]
Just a nit.

File: app-modules/ai/src/Exceptions/AiThreadLockedException.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    https://www.canyongbs.com or contact us via email at legal@canyongbs.com.
+
+</COPYRIGHT>
+*/
+
+namespace AdvisingApp\Ai\Exceptions;
+
+use Exception;
+
+class AiThreadLockedException extends Exception {}
```
[the remark this answers is not in the document]
Comment (danharrin) [reply app-modules/ai/src/Exceptions/AiThreadLockedException.php]: Done

File: app-modules/ai/src/Http/Controllers/RetryMessageController.php [hunk 2]
```diff
@@ -51,11 +52,16 @@ public function __invoke(RetryMessageRequest $request, AiThread $thread): JsonRe
                 $thread,
                 $request->validated('content'),
             );
+        } catch (AiThreadLockedException $exception) {
+            return response()->json([
+                'isThreadLocked' => true,
+                'message' => 'The assistant is currently undergoing maintenance.',
```
[the remark this answers is not in the document]
Comment (danharrin) [reply app-modules/ai/src/Http/Controllers/RetryMessageController.php]: Done

File: app-modules/ai/src/Http/Controllers/SendMessageController.php [hunk 3]
```diff
@@ -51,11 +52,16 @@ public function __invoke(SendMessageRequest $request, AiThread $thread): JsonRes
                 $thread,
                 $request->validated('content'),
             );
+        } catch (AiThreadLockedException $exception) {
+            return response()->json([
+                'isThreadLocked' => true,
+                'message' => 'The assistant is currently undergoing maintenance.',
```
[the remark this answers is not in the document]
Comment (danharrin) [reply app-modules/ai/src/Http/Controllers/SendMessageController.php]: Done

File: app-modules/ai/tests/Feature/Http/Controllers/RetryMessageControllerTest.php [hunk 4]
```diff
@@ -103,7 +104,37 @@
...
+        ->for(auth()->user())
+        ->create();
+
+    post(route('ai.threads.messages.retry', ['thread' => $thread]), [
+        'content' => AiMessage::factory()->make()->content,
+    ])
+        ->assertServiceUnavailable()
+        ->assertJson([
+            'isThreadLocked' => true,
+            'message' => 'The assistant is currently undergoing maintenance.',
```
[the remark this answers is not in the document]
Comment (danharrin) [reply app-modules/ai/tests/Feature/Http/Controllers/RetryMessageControllerTest.php]: Done

**Review by mxm1070 — APPROVED** [review]

**Review by Orrison — APPROVED** [review]

**Pull request merged**